hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -0,0 +1,360 @@
+
+	conversion, err := createTypeConversion(sourceEndpoint, destinationEndpoint)
+
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"trying to assign %s from %s",
+			destinationProperty.propertyName,
+			sourceProperty.propertyName)
+	}
+
+	return func(source dst.Expr, destination dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
+
+		var reader = astbuilder.Selector(source, string(sourceProperty.PropertyName()))
+		var writer = astbuilder.Selector(destination, string(destinationProperty.PropertyName()))
+
+		return conversion(reader, writer, ctx)
+	}, nil
+}
+
+// StorageTypeConversion generates the AST for a given conversion.
+// source is an expression to read the original value
+// destination is an expression to write the converted value
+// The parameters source and destination are funcs because AST fragments can't be reused, and in
+// some cases we need to reference source and destination multiple times in a single fragment.
+type StorageTypeConversion func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt
+
+// StorageTypeConversionFactory represents factory methods that can be used to create StorageTypeConversions
+// for a specific pair of types
+// source is the endpoint that will be read
+// destination is the endpoint that will be written
+type StorageTypeConversionFactory func(source *StorageConversionEndpoint, destination *StorageConversionEndpoint) StorageTypeConversion
+
+// A list of all known type conversion factory methods
+var typeConversionFactories []StorageTypeConversionFactory
+
+func init() {
+	typeConversionFactories = []StorageTypeConversionFactory{
+		assignPrimitiveTypeFromPrimitiveType,
+		assignOptionalPrimitiveTypeFromPrimitiveType,
+		assignPrimitiveTypeFromOptionalPrimitiveType,
+		assignArrayFromArray,
+		assignMapFromMap,
+	}
+}
+
+// createTypeConversion tries to create a type conversion between the two provided types, using
+// all of the available type conversion functions in priority order to do so.
+func createTypeConversion(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) (StorageTypeConversion, error) {
+	for _, f := range typeConversionFactories {
+		result := f(sourceEndpoint, destinationEndpoint)
+		if result != nil {
+			return result, nil
+		}
+	}
+
+	err := fmt.Errorf(
+		"no conversion found to assign %s from %s",
+		destinationEndpoint.name,
+		sourceEndpoint.name)
+
+	return nil, err
+}
+
+// assignPrimitiveTypeFromPrimitiveType will generate a direct assignment if both types have the
+// same underlying primitive type and have the same optionality
+//
+// <destination> = <source>
+//
+func assignPrimitiveTypeFromPrimitiveType(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) StorageTypeConversion {
+
+	srcOpt := AsOptionalType(sourceEndpoint.Type())
+	dstOpt := AsOptionalType(destinationEndpoint.Type())
+
+	if srcOpt != dstOpt {
```

Comment:
Should this be `.Equals`?

I do think you mentioned rewriting this function though.